### DISCUSSION TITLE
Working with 16-bits raw non-interleaved image data

### DISCUSSION THREAD

**Discussion by 4lch** [libvips/libvips#3863] [Q&A] [answered]
Hi and thanks in advance for your help !

I am working with a RGB+IR industrial line-scan camera that outputs 16-bits binary raw files that are simple pixel dumps (no header etc). However, the pixels are not interleaved in memory. The first quarter of the data represents all the red pixels, etc up to the last quarter that holds the IR data.

I am using Vips through NetVips, but I don't think it makes a huge difference in this case.

I have tried using the "rawload" function, using 8 bands (4 channels, with 2 bytes of data per pixel) and the correct width and height, then copying the image to set the number of bands and pixel format correctly (4 bands and Ushort format since the pixels are 16-bits unsigned integers). 

rawload seems to expect that the raw data is interleaved in memory, since I am getting the IR pixel data in the bottom 1/4 of the image on each band, rather than on its own band.

For reference, the Python code I currently use to load these files to a numpy array is the following, and works : 
``` Python
raw = np.fromfile(fn, np.uint16)
raw = raw.reshape((HEIGHT, WIDTH, 4))
```

Is there an elegant way to go about this that would not add too much overhead ?

Many thanks in advance and thank you for your time !

Best regards

**Comment by jcupitt** [comment, marked as answer]
Hi @4lch,

Probably the simplest would be to load as a one-band image, crop out the four parts, and use `bandjoin`. In untested python (just because it's easy):

```python
mono_image = pyvuips.Image.rawload("pixel-dump", 4096, 1024, 2).copy(bands=1, format="ushort")
bands = [mono_image.crop(0, y, 4096, 256) for y in range(0, 1024, 256)]
image = bands[0].bandjoin(bands[1:])
```

It should be reasonably quick. Have 1000 of those things, then use `arrayjoin(across=1)` to assemble them into a tall strip. You can crop out frames for adjustment, scaling to 8 bit and display, or just zoom over the whole thing.

vipsdisp has a display control bar for managing 16 -> 8 conversion, perhaps you saw? There's a win binary:

https://github.com/jcupitt/vipsdisp/releases

**Comment by 4lch** [comment]
Hi John and thanks a lot for your input here !

This worked great on test files I created for the purpose, but it turns out that the actual files are stranger than I thought...
In my previous example, the python code should actually have been this to load the RGB part of the image  from the pixel dump:
``` Python
raw = np.fromfile(fn, np.uint16)
raw = raw[:-len(raw)//4]
raw = raw.reshape((HEIGHT, WIDTH, 3))
```
As when doing that at first, I simply discarded the IR part. As it turns out, the raw file contains normal interleaved BGR data in the first 3/4 of it, and mono IR data in the last 1/4. I will look into pre-processing the data to save it as a color tiff and a mono IR tiff from the camera directly as this should make working with it a lot easier.

Thanks also for the `arrayjoin` tip ! It is awesome for this specific application to be able to manipulate the huge full strip image with such little overhead 😊

Thanks again for your help and your work on libvips, this is greatly appreciated !

Best regards

**Reply by jcupitt** [reply]
Wow OK that is a bit crazy. I've worked on several camera projects and the hardware people always do something odd.

You could load them separately, eg.:

```python
rgb_image = pyvips.Image.rawload("pixel-dump", 4096, 1024, 6) \
    .copy(bands=3, format="ushort")
ir_image = pyvuips.Image.rawload("pixel-dump", 4096, 1024, 2) \
    .copy(bands=1, format="ushort") \
    .crop(0, 768, 4096, 256)
image = rgb_image.bandjoin(ir_image)
```

The downside is that you'll now need 2,000 files open at once, which is the hard limit on Windows (no more than 2000 files per process). You'd probably need to implement your own caching and file handle reuse thing.

You could keep the IR and RGB paths separate, maybe that would be simpler? I can imagine they might need different handling anyway, for example the IR image is likely to look softer and will need different noise filtering.

**Reply by 4lch** [reply]
Thanks again for your input John !
This sounds very sensible to work with these files, I don't know what the hardware guys were thinking with this format 😁

Indeed, it sounds like we would be flying a little close to the sun with this amount of open files, as we expect to get at most 1200-1500 chunks per full strip. RGB data and IR are used in different contexts and at different times as you mentioned so the best would surely be to decorrelate them entirely. After speaking with the camera manufacturer, it does look like we could split the data to two buffers and save each as separate raws.

Thanks again so much for your help and time, it's also nice that you have experience working with such cameras and scanners ! I am glad to have found Vips as it should make things a lot easier for this application :)
